File: src/rust/src/bin/day_one.rs
use std::fs::read_to_string;

#[derive(Debug)]
struct Rotation {
    direction: String,
    amount: i32,
}

fn main() {
    let testing = false;
    let mut file_path = "../../input/1.in";
    if testing {
        file_path = "../../input/1.test"
    }
    let input = read_to_string(file_path).expect("file read failed");
    let ans = part_one(&input);
    println!("part one {ans}");
    let ans = part_two(&input);
    println!("part two {ans}");
}

fn part_one(input: &String) -> i32 {
    let values: Vec<Rotation> = input.lines().map(|l| parse(l.to_string())).collect();
    let mut last = 50;
    let mut zero_count = 0;
    for r in values {
        let new = rotate(last, &r);
        if new == 0 {
            zero_count += 1;
        }
        last = new;
    }
    zero_count
}

fn part_two(input: &String) -> i32 {
    let values: Vec<Rotation> = input.lines().map(|l| parse(l.to_string())).collect();
    let mut last = 50;
    let mut zero_count = 0;
    for r in values {
        let (new, zeros) = rotate_by_step(last, &r);
        zero_count += zeros;
        last = new;
    }
    zero_count
}

fn rotate(current: i32, rotation: &Rotation) -> i32 {
    let ans;
    match rotation.direction.as_str() {
        "L" => ans = (current - rotation.amount + 100) % 100,
        "R" => ans = (current + rotation.amount) % 100,
        _ => panic!("oh no"),
    }
    ans
}

fn rotate_by_step(current: i32, rotation: &Rotation) -> (i32, i32) {
    let mut ans = 0;
    let mut zeroes = 0;
    let mut last = current;
    for _i in 0..rotation.amount {
        match rotation.direction.as_str() {
            "L" => ans = (last - 1 + 100) % 100,
            "R" => ans = (last + 1) % 100,
            _ => panic!("oh no"),
        }
        last = ans;
        if ans == 0 {
            zeroes += 1;
        }
    }
    (ans, zeroes)
}

fn parse(val: String) -> Rotation {
    let direction: String = val[0..1].to_string();
    let amount = val[1..].parse().unwrap();
    Rotation { direction, amount }
}

File: src/node/day_one.js
const fs = require('node:fs');

fs.readFile('../../input/1.in', 'utf8', (err, data) => {
    if (err) {
        console.error(err);
        return;
    }
    let values = data.trim().split('\n');
    let part_one_ans = 0;
    let last = 50;
    let next = 0;
    values.forEach(v => {
        let rotation = parse(v);
        next = rotate(last, rotation);
        if (next == 0) {
            part_one_ans += 1;
        }
        last = next;
    });
    console.log("part one:", part_one_ans);
});


function rotate(current, rotation) {
    let res = 0;
    if (rotation.direction == "L") {
         res = (current - rotation.amount + 100) % 100;
    } else {
        res = (current + rotation.amount) % 100;
    }
    return res;
}

function parse(entry) {
    return {"direction":entry[0], "amount":+entry.slice(1)}
}